cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -1470,12 +1470,16 @@
 (rule (lower (icmp (IntCC.SignedLessThan) a @ (value_type (vec128 ty)) b))
       (x64_pcmpgt ty b a))
 (rule (lower (icmp (IntCC.UnsignedGreaterThan) a @ (value_type (vec128 ty)) b))
-      (let ((max Xmm (x64_pmaxu ty a b))
+      ;; N.B.: we must manually prevent load coalescing of these operands; the
+      ;; register allocator gets confused otherwise. TODO:
+      ;; https://github.com/bytecodealliance/wasmtime/issues/3953.
+      (let ((max Xmm (x64_pmaxu ty (put_in_xmm a) (put_in_xmm b)))
             (eq Xmm (x64_pcmpeq ty max b))
```

Comment:
I haven't been following the implicit conversions recently but this feels sort of weird to me where previously when I was working on ISLE I tried to make sure that `put_in_xmm` was only called once-per-value, but here you're explicitly calling `put_in_xmm b` above but here it's implicitly being converted with what I presume is also `put_in_xmm`. I don't personally know the effect of calling `put_in_xmm` twice and whether it has some sort of cache to repeatedly return the same thing.

Reply:
There's actually a whole section of our documentation answering this question! [Here](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/docs/isle-integration.md#implicit-type-conversions-and-side-effects) we added some justification for this decision and a criterion for when it is acceptable: implicit conversions can have side-effects if they are *idempotent*.

Specifically, `put_in_{reg,xmm,gpr}` will make a note that the value is required (so when the upward pass reaches the producer, it's actually lowered), and return the vreg already assigned for this SSA value. Calling it multiple times will return the same vreg.

On balance, making `put_in_reg` implicit makes lowering patterns easier to write, I think; the real issue here is the odd condition that we can't use memory operands in compares. This is (as @abrown notes in sibling issue) something we should fix, for sure.

Reply:
This thread is resolved by the new `let` statements for `xmm_a` and `xmm_b` in c8d8556. I'll leave it unresolved, though, for future readers who may want to understand what @cfallin just explained.